### PR TITLE
[SPARK-27783][SQL] Add customizable hint error handler

### PR DESCRIPTION
**Pull request by maryannxue**
## What changes were proposed in this pull request?

Added an interface for handling hint errors, with a default implementation class that logs warnings in the callbacks.

## How was this patch tested?

Passed existing tests.

### CODE REVIEW

**Comment by maryannxue** [comment]
retest this please

**Comment by maryannxue** [comment]
cc @maropu @gatorsmile @cloud-fan Could you guys please take a look?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/hints.scala [hunk 1]
```diff
@@ -157,3 +158,38 @@ case object SHUFFLE_REPLICATE_NL extends JoinStrategyHint {
...
+
+  /**
+   * Callback for relation names specified in a hint that cannot be associated with any relation
+   * in the current scope.
+   * @param name the hint name
+   * @param parameters the hint parameters
+   * @param invalidRelations the set of relation names that cannot be associated
+   */
+  def handleHintRelationsNotFound(
+    name: String, parameters: Seq[Any], invalidRelations: Set[String]): Unit
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/hints.scala]: nit: 4 space indentation

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala [hunk 2]
```diff
@@ -74,11 +76,13 @@ object ResolveHints {
           case ResolvedHint(u: UnresolvedRelation, hint)
               if relations.exists(resolver(_, u.tableIdentifier.table)) =>
             relations.remove(u.tableIdentifier.table)
-            ResolvedHint(u, createHintInfo(hintName).merge(hint, handleOverriddenHintInfo))
+            ResolvedHint(u,
+              createHintInfo(hintName).merge(hint, hintErrorHandler.handleHintOverridden))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala]: nit: can we just pass the `hintErrorHandler` to `merge`? That seems more flexible.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala [hunk 3]
```diff
@@ -130,10 +130,7 @@ object ResolveHints {
           relationNames.foreach(relationNameSet.add)
 
           val applied = applyJoinStrategyHint(h.child, relationNameSet, h.name)
-          relationNameSet.foreach { n =>
-            logWarning(s"Count not find relation '$n' for join strategy hint " +
-              s"'${h.name}${relationNames.mkString("(", ", ", ")")}'.")
-          }
+          hintErrorHandler.handleHintRelationsNotFound(h.name, h.parameters, relationNameSet.toSet)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala]: This is specific to join hints right? Maybe we should rename `handleHintRelationsNotFound` to something join specific.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/EliminateResolvedHint.scala [hunk 4]
```diff
@@ -37,7 +41,7 @@ object EliminateResolvedHint extends Rule[LogicalPlan] {
     }
     pulledUp.transformUp {
       case h: ResolvedHint =>
-        handleInvalidHintInfo(h.hints)
+        hintErrorHandler.handleJoinNotFoundForJoinHint(h.hints)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/EliminateResolvedHint.scala]: This seems not join specific, maybe `handleInvalidHintInfo` is a better name.

File: sql/core/src/test/scala/org/apache/spark/sql/JoinHintSuite.scala [hunk 5]
```diff
@@ -281,7 +281,7 @@ class JoinHintSuite extends PlanTest with SharedSQLContext {
           JoinHint(
             Some(HintInfo(strategy = Some(SHUFFLE_HASH))),
             Some(HintInfo(strategy = Some(SHUFFLE_HASH)))) :: Nil,
-        msgNoHintRelationFound("d", "merge(b, d)") ::
+        msgNoHintRelationFound("d", "merge('b, 'd)") ::
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinHintSuite.scala]: it's weird to see `'b` representing a table name, `'b` usually represents a column.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala (hunk 2 above)
Comment (maryannxue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala]: Yeah, that makes sense.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala (hunk 3 above)
Comment (maryannxue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala]: At this moment we don't have many other hints, but I'd not say it's join specific. Like in other systems, there's index hint that can be and should be bound to a relation.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/EliminateResolvedHint.scala (hunk 4 above)
Comment (maryannxue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/EliminateResolvedHint.scala]: No. The original name is not accurate. This rule basically extract hints nodes from child relations of a join and pull them up to the Join node. If we come up with other "invalid hints" later, we can have other error handler callbacks specific to those invalid hints.

File: sql/core/src/test/scala/org/apache/spark/sql/JoinHintSuite.scala (hunk 5 above)
Comment (maryannxue) [reply sql/core/src/test/scala/org/apache/spark/sql/JoinHintSuite.scala]: That's the default output from the hint parameters. I simply called `toString` for all the parameters.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/EliminateResolvedHint.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/EliminateResolvedHint.scala]: This rule is called `EliminateResolvedHint`, is it possible that we add other hints and also report hint target not found issue here?

Comment (maryannxue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/EliminateResolvedHint.scala]: The hope is we'll be able to recognize the specific hint error in this rule if anything else. Say if a XYZ-hint has to be associated with XYZ-node, we should another error assigned to it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HintErrorLogger.scala [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+  override def hintNotRecognized(name: String, parameters: Seq[Any]): Unit = {
+    logWarning(s"Unrecognized hint: $name${parameters.mkString("(", ", ", ")")}")
+  }
+
+  override def hintRelationsNotFound(
+      name: String, parameters: Seq[Any], invalidRelations: Set[String]): Unit = {
+    invalidRelations.foreach { n =>
+      logWarning(s"Count not find relation '$n' for join strategy hint " +
+        s"'$name${parameters.mkString("(", ", ", ")")}'.")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HintErrorLogger.scala]: can we keep the previous behavior? https://github.com/apache/spark/pull/24653/files#diff-746a6d090224c7cfbe15daa27fa27408L135

`${h.name}${relationNames.mkString("(", ", ", ")")}`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala [hunk 7]
```diff
@@ -74,11 +76,13 @@ object ResolveHints {
           case ResolvedHint(u: UnresolvedRelation, hint)
               if relations.exists(resolver(_, u.tableIdentifier.table)) =>
             relations.remove(u.tableIdentifier.table)
-            ResolvedHint(u, createHintInfo(hintName).merge(hint, handleOverriddenHintInfo))
+            ResolvedHint(u,
+              createHintInfo(hintName).merge(hint, hintErrorHandler))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala]: nit: unnecessary to split the line

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala [hunk 8]
```diff
@@ -74,11 +76,13 @@ object ResolveHints {
...
           case ResolvedHint(r: SubqueryAlias, hint)
               if relations.exists(resolver(_, r.alias)) =>
             relations.remove(r.alias)
-            ResolvedHint(r, createHintInfo(hintName).merge(hint, handleOverriddenHintInfo))
+            ResolvedHint(r,
+              createHintInfo(hintName).merge(hint, hintErrorHandler))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HintErrorLogger.scala [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    }
+  }
+
+  override def joinNotFoundForJoinHint(hint: HintInfo): Unit = {
+    logWarning(s"A join hint $hint is specified but it is not part of a join relation.")
+  }
+
+  override def hintOverridden(hint: HintInfo): Unit = {
+    logWarning(s"Join hint $hint is overridden by another hint and will not take effect.")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HintErrorLogger.scala]: shall we make the method name specific to join hint? or make the error log general?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HintErrorLogger.scala [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+object HintErrorLogger extends HintErrorHandler with Logging {
+
+  override def hintNotRecognized(name: String, parameters: Seq[Any]): Unit = {
+    logWarning(s"Unrecognized hint: $name${parameters.mkString("(", ", ", ")")}")
+  }
+
+  override def hintRelationsNotFound(
+      name: String, parameters: Seq[Any], invalidRelations: Set[String]): Unit = {
+    invalidRelations.foreach { n =>
+      logWarning(s"Count not find relation '$n' for join strategy hint " +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HintErrorLogger.scala]: is this specific to join?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HintErrorLogger.scala (hunk 6 above)
Comment (maryannxue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HintErrorLogger.scala]: I wanted to print the original hint with all parameters, but it turns out the relation names aren't printed nicely. Is there a way out of this?

**Review by gatorsmile — COMMENTED** [review]
LGTM

Thanks! Merged to master.

**Pull request closed without merging**